flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetSingleRowJoin.scala

```diff
@@ -144,21 +150,46 @@ class DataSetSingleRowJoin(
 
     val condition = codeGenerator.generateExpression(joinCondition)
 
-    val joinMethodBody = s"""
-                  |${condition.code}
-                  |if (${condition.resultTerm}) {
-                  |  ${conversion.code}
-                  |  ${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
-                  |}
-                  |""".stripMargin
+    val joinMethodBody =
+      if (joinType == JoinRelType.INNER) {
+        s"""
+         |${condition.code}
+         |if (${condition.resultTerm}) {
+         |  ${conversion.code}
+         |  ${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
+         |}
+         |""".stripMargin
+    } else {
+        val singleNode =
+          if (leftIsSingle) {
+            leftNode
+          }
+          else {
+            rightNode
+          }
+        val notSuitedToCondition = singleNode
+          .getRowType
+          .getFieldList
+          .map(field => getRowType.getFieldNames.indexOf(field.getName))
+          .map(i => s"${conversion.resultTerm}.setField($i,null);")
```

Comment:
This only works if the result term is a `Row`. 

It think a better approach is to generate a dedicated result expression for the null case based on `CodeGenerator.generateResultExpression()`, where the field expressions are the outer fields and the null expressions. Then we could use the current result expression if the condition is met and the result expression with null values if that is not the case.

We would need to extend the `CodeGenerator` with methods for `generateFieldAccesses(boolean firstInput)` (see `generateConverterResultExpression()`) and `generateNullValues(boolean firstInput)` which generates expressions for null values.